crml/transaction-payment/src/lib.rs

```diff
@@ -161,6 +161,29 @@ impl<T: Trait> Module<T> {
 	}
 }
 
+#[derive(Debug)]
+pub struct FeeParts {
+	/// The length fee
+	length_fee: u128,
+	/// The weight fee after conversion
+	weight_fee: u128,
+	/// The len + fee after adjustment factor is applied
+	adjusted_fee: u128,
+	/// The final fee
+	final_fee: u128,
```

Comment:
If we remove this field, we have prevented from a possible discrepancy between the other fields and this one. Instead, we can take the lazy evaluation approach and calculate final_fee only when final_fee is called.

Reply:
In fact, I would want to suggest that we only keep the actual building blocks of a fee, which is the length of the encoded tx, its weight factor and the adjustment factor. Everything else like the fee parts should be methods for calculating the elements. Final fee can be a function (not a method) which would call the other methods to calculate the Final fee. Something like the following pseudo-code: 
`
pub struct Fee {
   const BASE: Balance = x,
   const PER_BYTE: Balance = y, 
   encoded_tx_len: usize,
   tx_weight: Balance,
   network_usage_adjustment: u128
}

impl Fee {
   fn length_fee(&self): Balance { ... }
   fn weight_fee(&self): Balance { ... }
   fn adjusted_fee(&Self x): Balance { (x.length_fee()+x.weight_fee())*x.network_usage_adjustment }
   pub fn final_fee(&Self x): Balance { x.BASE + x.adjusted_fee() }
}
}
`

Reply:
The above approach has got the following characteristics:

- It's a lazy evaluation approach. 
- Each private method does only one single cohesive job.
- Only functions are allowed to orchestrate methods for any additional service. Methods are not going to call each other and make an unnecessary intermingled structure. The benefit would be more obvious if later someone needs to add a thread lock to methods for example. With this approach, locks not going to get intertwined or reentered.
- The generic type of Balance is preferred to u128 as it integrates better with the rest of cennznet.
- For each member variable, the preferred type is used so the compiler can get possible type mistakes by type checking.

Reply:
If we need a cache feature too, then we can do:
```
pub struct CachedFee<Balance> {
   fee: Fee<Balance>,
   update_needed: bool,
   cached_fee: Balance
}

impl<Balance> CachedFee<Balance> {
   pub fn set_tx_len(&mut self, x) {
      self.fee.encoded_tx_len = x;
      self.update_needed = true; 
   }
  
   pub set..(&mut self, x) {
      self.fee... = x;
      self.update_needed = true;
   }
   pub fn final_fee(&mut self) { 
      if (self.update_needed) {
          self.cached_fee = self.fee.final_fee();
          self.update_needed = false;
      }
  }
}
```